### PR TITLE
Lab consistency changes, Fixes AB#3043333

### PR DESCRIPTION
**Pull request by fadidurah**
Changes to fix some ui tests involving 1st party apps, and changes to improve lab api reliability

changes:
- Add a retry to getAccessToken calls by default in LabApiAuthenticationClient
- Reduce temp user api call timeout to account for introducing more retries
- ~~Add a retry in ApiClient~~ Removed this, retrying at this level will through exception, okhttp calls cannot be executed twice (https://square.github.io/okhttp/2.x/okhttp/index.html?com/squareup/okhttp/Call.html). Instead, I added a retry in LabClient for user fetch.
- Add some general ui handling for outlook and teams

[AB#3043333](https://identitydivision.visualstudio.com/fac9d424-53d2-45c0-91b5-ef6ba7a6bf26/_workitems/edit/3043333)

### CODE REVIEW

File: LabApiUtilities/src/main/com/microsoft/identity/labapi/utilities/authentication/LabApiAuthenticationClient.java [hunk 1]
```diff
@@ -72,6 +76,45 @@ public LabApiAuthenticationClient(@NonNull final String labSecret, final String
...
+        // Do this in a loop, if we get an exception or null result, try again
+        for (int i = 0; i < retries; i++) {
+            System.out.printf(Locale.ENGLISH, "getAccessToken attempt #%d%n", (i + 1));
+
+            try {
+                final String result = getAccessTokenInternal();
+                if (result != null) {
+                    return result;
+                }
+            } catch (LabApiException labApiException) {
```
Comment (shahzaibj) [inline LabApiUtilities/src/main/com/microsoft/identity/labapi/utilities/authentication/LabApiAuthenticationClient.java]:
```suggestion
            } catch (final LabApiException labApiException) {
```

File: LabApiUtilities/src/main/com/microsoft/identity/labapi/utilities/authentication/LabApiAuthenticationClient.java [hunk 2]
```diff
@@ -72,6 +76,45 @@ public LabApiAuthenticationClient(@NonNull final String labSecret, final String
...
     @Override
     public String getAccessToken() throws LabApiException {
+        return getAccessToken(DEFAULT_ACCESS_TOKEN_RETRIES);
+    }
+
+    public String getAccessToken(final int retries) throws LabApiException {
+
+        // Do this in a loop, if we get an exception or null result, try again
+        for (int i = 0; i < retries; i++) {
+            System.out.printf(Locale.ENGLISH, "getAccessToken attempt #%d%n", (i + 1));
```
Comment (shahzaibj) [inline LabApiUtilities/src/main/com/microsoft/identity/labapi/utilities/authentication/LabApiAuthenticationClient.java]: Could consider starting the loop from `i = 1` and until `i <= retries` then you wouldn't need to add/subtract from i when printing

File: labapi/src/main/java/com/microsoft/identity/internal/test/labapi/ApiClient.java [hunk 3]
```diff
@@ -854,8 +855,25 @@ public <T> ApiResponse<T> execute(Call call, Type returnType) throws ApiExceptio
             Response response = call.execute();
             T data = handleResponse(response, returnType);
             return new ApiResponse<T>(response.code(), response.headers().toMultimap(), data);
-        } catch (IOException e) {
-            throw new ApiException(e);
+        } catch (IOException exception) {
```
Comment (shahzaibj) [inline labapi/src/main/java/com/microsoft/identity/internal/test/labapi/ApiClient.java]:
```suggestion
        } catch (final IOException exception) {
```

File: labapi/src/main/java/com/microsoft/identity/internal/test/labapi/ApiClient.java [hunk 4]
```diff
@@ -854,8 +855,25 @@ public <T> ApiResponse<T> execute(Call call, Type returnType) throws ApiExceptio
...
+            // If we receive an IO Exception, we should retry once, to avoid inconsistent
+            // network causing test failures. The timeout has been reduced to allow this.
+
+            try {
+                // Wait for a bit
+                try {
+                    Thread.sleep(LAB_API_RETRY_WAIT);
+                } catch (final InterruptedException e2) {
+                    e2.printStackTrace();
```
Comment (shahzaibj) [inline labapi/src/main/java/com/microsoft/identity/internal/test/labapi/ApiClient.java]:
```suggestion
                } catch (final InterruptedException exception) {
                    exception.printStackTrace();
```

File: labapi/src/main/java/com/microsoft/identity/internal/test/labapi/ApiClient.java [hunk 5]
```diff
@@ -854,8 +855,25 @@ public <T> ApiResponse<T> execute(Call call, Type returnType) throws ApiExceptio
...
+                }
+                Response response = call.execute();
+                T data = handleResponse(response, returnType);
```
Comment (shahzaibj) [inline labapi/src/main/java/com/microsoft/identity/internal/test/labapi/ApiClient.java]:
```suggestion
                final Response response = call.execute();
                final T data = handleResponse(response, returnType);
```

File: labapi/src/main/java/com/microsoft/identity/internal/test/labapi/ApiClient.java [hunk 6]
```diff
@@ -854,8 +855,25 @@ public <T> ApiResponse<T> execute(Call call, Type returnType) throws ApiExceptio
...
+                return new ApiResponse<T>(response.code(), response.headers().toMultimap(), data);
+            } catch (IOException secondException) {
```
Comment (shahzaibj) [inline labapi/src/main/java/com/microsoft/identity/internal/test/labapi/ApiClient.java]:
```suggestion
            } catch (final IOException secondException) {
```

Comment (shahzaibj) [reply labapi/src/main/java/com/microsoft/identity/internal/test/labapi/ApiClient.java]: I think we can just call it exception vs secondException

File: labapi/src/main/java/com/microsoft/identity/internal/test/labapi/ApiClient.java [hunk 7]
```diff
@@ -854,8 +855,25 @@ public <T> ApiResponse<T> execute(Call call, Type returnType) throws ApiExceptio
...
+                // If we get another IO Exception, we should fail. Most likely another retry
+                // will not resolve the issue.
+                throw new ApiException(secondException);
+            }
```
Comment (shahzaibj) [inline labapi/src/main/java/com/microsoft/identity/internal/test/labapi/ApiClient.java]: It seems we are just repeating the code in 855-857.

Instead could consider using recursion and with an n value that terminates the recursion once n reaches 2 in this case i.e. 2 io exceptions

File: LabApiUtilities/src/main/com/microsoft/identity/labapi/utilities/authentication/LabApiAuthenticationClient.java [hunk 8]
```diff
@@ -72,6 +76,45 @@ public LabApiAuthenticationClient(@NonNull final String labSecret, final String
 
     @Override
     public String getAccessToken() throws LabApiException {
+        return getAccessToken(DEFAULT_ACCESS_TOKEN_RETRIES);
+    }
+
+    public String getAccessToken(final int retries) throws LabApiException {
```
Comment (shahzaibj) [inline LabApiUtilities/src/main/com/microsoft/identity/labapi/utilities/authentication/LabApiAuthenticationClient.java:88]: javadoc for public methods

File: LabApiUtilities/src/main/com/microsoft/identity/labapi/utilities/authentication/LabApiAuthenticationClient.java (hunk 2 above)
Comment (fadidurah) [reply LabApiUtilities/src/main/com/microsoft/identity/labapi/utilities/authentication/LabApiAuthenticationClient.java]: Updated to reflect this

**Review by melissaahn — APPROVED** [review]

**Review by shahzaibj — APPROVED** [review]

**Pull request merged**